### PR TITLE
Re-implement MempoolCapacity in terms of bytes

### LINKED ISSUE #974: Mempool: reimplement MempoolCapacity in terms of bytes
Once PR #890 (issue #621) is merged, the mempool will support the ability to be limited by a specified `MempoolCapacity`. However, this `MempoolCapacity` is defined in terms of number of transactions. It would be more useful for this to be defined in terms of bytes.

From @dcoutts (https://github.com/input-output-hk/ouroboros-network/pull/890#discussion_r311494572):
> The only change I can imagine here (either now or in future) is that it might make sense for us to switch from the size being in number of txs, to being the total size in bytes of the txs in the pool.

### PR DESCRIPTION
**Pull request by intricate**
Closes #974

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs [hunk 1]
```diff
@@ -268,26 +270,26 @@ implAddTxs mpEnv accum txs = assert (all txInvariant txs) $ do
...
     -- We record the time at which the transactions were added to the mempool
     -- so we can use it in our performance measurements.
     time <- getMonotonicTime
 
-    traceBatch TraceMempoolRemoveTxs   mempoolSize (map fst removed) time
-    traceBatch TraceMempoolAddTxs      mempoolSize accepted          time
-    traceBatch TraceMempoolRejectedTxs mempoolSize rejected          time
+    traceBatch TraceMempoolRemoveTxs   numTxsMempool (map fst removed) time
+    traceBatch TraceMempoolAddTxs      numTxsMempool accepted          time
+    traceBatch TraceMempoolRejectedTxs numTxsMempool rejected          time
```
Comment (intricate) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs]: Should probably trace the entire `MempoolSize` instead

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 2]
```diff
@@ -774,19 +850,25 @@ mempoolCapTestExpectedTrace mcts = go chunks
...
-  , mctEnvRemovedTxs :: StrictTVar m Word
+  { mctEnvAddedBytes           :: StrictTVar m Word32
+    -- ^ The total number of bytes worth of transactions which have been added
+    -- to the mempool.
+  , mctEnvRemovedTxs           :: StrictTVar m Word
     -- ^ The number of transactions which have been removed from the mempool
     -- and added to the ledger.
+  , mctEnvBytesToBeAddedForCap :: StrictTVar m Word32
+    -- ^ The number of bytes that need to be added to the mempool to reach
+    -- full capacity.
```
Comment (intricate) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: I don't think this adequately explains the overall purpose of this field. Will add better doc.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 3]
```diff
@@ -894,7 +976,7 @@ prop_Mempool_idx_consistency (Actions actions) =
     emptyTestSetup = TestSetup
       { testLedgerState = testInitLedger
       , testInitialTxs  = []
-      , testMempoolCap  = MempoolCapacity 1000
+      , testMempoolCap  = MempoolCapacity 1000000
```
Comment (intricate) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Perhaps it'd be a good idea to instead define `MempoolCapacity` as a sum type with a `NoMempoolCapacity` constructor that we can utilize in tests that don't really care about capacity.

Comment (mrBliss) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: We can just use `maxBound` in the tests, if that's ever an issue in the tests, then there's something wrong with the tests (or the mock transaction sizes).

Comment (mrBliss) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Maybe it's good time to address #745 too?

Comment (intricate) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Yeah, for sure. I'll address #745 in a follow up PR.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 4]
```diff
@@ -715,22 +745,66 @@ instance Arbitrary MempoolCapTestSetup where
...
+    go :: Word32
+        -> [GenTx TestBlock]
+        -> [GenTx TestBlock]
+        -> ([GenTx TestBlock], [GenTx TestBlock])
+    go _ accTxs [] = (accTxs, [])
+    go accByteSize accTxs (t:ts)
+      | (txSize t + accByteSize) <= mpCap =
+        go (accByteSize + txSize t) (accTxs ++ [t]) ts
+      | otherwise = (accTxs, t:ts)
+
```
[the remark this answers is not in the document]
Comment (intricate) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Related: https://github.com/input-output-hk/ouroboros-network/pull/1346 (PR for #686)

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs [hunk 5]
```diff
@@ -90,9 +91,9 @@ data LedgerInterface m blk = LedgerInterface
   { getCurrentLedgerState :: STM m (LedgerState blk)
   }
 
--- | Represents the maximum number of transactions that a 'Mempool' can
--- contain.
-newtype MempoolCapacity = MempoolCapacity Word
+-- | Represents the maximum number of bytes worth of transactions that a
+-- 'Mempool' can contain.
+newtype MempoolCapacity = MempoolCapacity Word32
```
[the remark this answers is not in the document]
Comment (intricate) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs]: Thanks, I'll raise a PR for `cardano-byron-proxy` once this is approved and merged!

I didn't find any usages of `MempoolCapacity` in other projects such as `cardano-node` and I think I've gotten all of them in this repo.

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs [hunk 6]
```diff
@@ -366,12 +368,16 @@ implAddTxs mpEnv accum txs = assert (all txInvariant txs) $ do
       -- (those not yet validated due to the mempool capacity being reached).
     validateNewUntilMempoolFull []      vr = pure (vr, [])
     validateNewUntilMempoolFull (t:ts)  vr = do
-      mempoolSize <- getMempoolSize mpEnv
+      MempoolSize { msNumBytes } <- getMempoolSize mpEnv
+      let newValidSizeInBytes = Foldable.foldl'
```
[the remark this answers is not in the document]
Comment (intricate) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs]: So I added a function, `txsToMempoolSize` (that utilizes the `toMempoolSize` function you suggested in https://github.com/input-output-hk/ouroboros-network/pull/1342#discussion_r356496229), and I'm now using it here as well. Mind giving me your me your opinion on the change?

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 7]
```diff
@@ -260,7 +260,7 @@ mkNodeArgs registry cfg initState tracers btime chainDB = NodeArgs
     , blockFetchSize      = nodeBlockFetchSize
     , blockMatchesHeader  = nodeBlockMatchesHeader
     , maxUnackTxs         = 100 -- TODO
-    , mempoolCap          = MempoolCapacity 1000
+    , mempoolCap          = MempoolCapacity 1000000 -- TODO
```
Comment (intricate) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: @mrBliss: Eventually this should be configurable, but any idea what a good `MempoolCapacity` would be?

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs [hunk 8]
```diff
@@ -366,12 +368,24 @@ implAddTxs mpEnv accum txs = assert (all txInvariant txs) $ do
...
+
+      -- Determine what the mempool size would be if we were to commit the new
+      -- transactions we've validated thus far and also the next transaction
+      -- to validate, 't'.
+      -- If this value is greater than the 'MempoolCapacity', then we know not
+      -- to continue validating at this time.
+      let MempoolSize _ nextMsNumBytes =
+               mpsz
+            <> txsToMempoolSize (vrNewValid vr)
+            <> txsToMempoolSize [t]
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs]: Hmm, my suggestion was not that good after all. What about the following:
```haskell
let newTxsBytes    = sum (txSize <$> vrNewValid vr) + txSize t
    newSizeInBytes = curSizeInBytes + newTxsBytes
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs (hunk 7 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: @dcoutts: can you suggest a good default capacity in bytes for the Mempool of the node?

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 9]
```diff
@@ -402,29 +417,51 @@ ppTestSetup TestSetup { testLedgerState
...
+txsToMempoolSize :: [GenTx TestBlock] -> MempoolSize
+txsToMempoolSize = foldl'
+    (\MempoolSize { msNumTxs, msNumBytes } tx ->
+      MempoolSize
+        { msNumTxs   = msNumTxs + 1
+        , msNumBytes = msNumBytes + txSize tx
+        }
+    )
+    MempoolSize { msNumTxs = 0, msNumBytes = 0 }
+
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Use `txsToMempoolSize` here

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 10]
```diff
@@ -715,22 +757,70 @@ instance Arbitrary MempoolCapTestSetup where
...
+    pure MempoolCapTestSetup { mctsTestSetup = testSetup'
                              , mctsValidTxs  = vtxs
-                             , mctsCapacity  = MempoolCapacity (fromIntegral capacity)
+                             , mctsCapacity  = MempoolCapacity capacity
                              }
 
+-- | The 'MempoolSize' for a mempool consisting of no transactions.
+zeroMempoolSize :: MempoolSize
+zeroMempoolSize = MempoolSize { msNumTxs = 0, msNumBytes = 0 }
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: This is `mempty`

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 11]
```diff
@@ -715,22 +757,70 @@ instance Arbitrary MempoolCapTestSetup where
...
+splitTxsUntilCap txs (MempoolCapacity mpCap) = go 0 [] txs
+  where
+    go :: Word32
+        -> [GenTx TestBlock]
+        -> [GenTx TestBlock]
+        -> ([GenTx TestBlock], [GenTx TestBlock])
+    go _ accTxs [] = (accTxs, [])
+    go accByteSize accTxs (t:ts)
+      | (txSize t + accByteSize) <= mpCap =
+        go (accByteSize + txSize t) (accTxs ++ [t]) ts
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Actually, this is *O(n^2)*!

Instead of `accTxs ++ [t]`, which is *O(n)*, use `(t:accTxs)` (*O(1)*) and `reverse` (*O(n)*) the accumulator when you return it (on L818 and L822). I know this is just test code, its performance doesn't really matter in practice, and you'll update it in #1346, but still :slightly_smiling_face:

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 12]
```diff
@@ -745,23 +835,22 @@ mempoolCapTestExpectedTrace mcts = go chunks
...
-    chunksOf _ [] = []
-    chunksOf n xs
-      | n > 0     = take n xs : chunksOf n (drop n xs)
-      | otherwise = error $  "mempoolCapTestExpectedTrace.chunksOf: "
-                          <> "n is less than or equal to 0"
+    buildChunks :: [[GenTx TestBlock]] -> [GenTx TestBlock] -> [[GenTx TestBlock]]
+    buildChunks acc []     = acc
+    buildChunks acc remTxs =
+      let (ts, ts') = splitTxsUntilCap remTxs (MempoolCapacity mempoolCap)
+      in buildChunks (acc ++ [ts]) ts'
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Same here: `ts : acc` + `reverse`

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 13]
```diff
@@ -774,19 +863,25 @@ mempoolCapTestExpectedTrace mcts = go chunks
 -- | A data type containing 'StrictTVar's by which the two threads spawned by
 -- 'prop_Mempool_Capacity' can coordinate with each other.
 data MempoolCapTestEnv m = MempoolCapTestEnv
-  { mctEnvAddedTxs   :: StrictTVar m Word
-    -- ^ The number of transactions which have been added to the mempool.
-  , mctEnvRemovedTxs :: StrictTVar m Word
+  { mctEnvAddedTxs             :: StrictTVar m Word32
+    -- ^ The total number of transactions which have been added to the
+    -- mempool.
+  , mctEnvRemovedTxs           :: StrictTVar m Word
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: This should be the same type as `mctEnvAddedTxs`. I suggest you change them both to `Int` as it is just a count used in the tests, so it'll get rid of some `fromIntegral`s.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 14]
```diff
@@ -715,22 +757,70 @@ instance Arbitrary MempoolCapTestSetup where
...
+        capacityMinBound = maximum txSizes
+        -- The maximum mempool capacity should be the sum of all of the
+        -- transaction sizes.
+        capacityMaxBound = sum txSizes
+    capacity <- choose
+      ( capacityMinBound
+      , capacityMaxBound
+      )
+
+    -- As mentioned above, we need to  adjust the originally generated mempool
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]:
```suggestion
    -- As mentioned above, we need to adjust the originally generated mempool
```

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 15]
```diff
@@ -745,23 +835,22 @@ mempoolCapTestExpectedTrace mcts = go chunks
...
+      let (ts, ts') = splitTxsUntilCap remTxs (MempoolCapacity mempoolCap)
+      in buildChunks (acc ++ [ts]) ts'
 
     chunks :: [[GenTx TestBlock]]
-    chunks = chunksOf (fromIntegral mempoolCap) txs
+    chunks = buildChunks [] txs
 
     go :: [[GenTx TestBlock]] -> [TraceEventMempool TestBlock]
     go []         = []
     go (chunk:cs) = chunkExpectedTrace chunk ++ go cs
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Replace this with `concatMap chunkExpectedTrace` and inline it in L828.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 16]
```diff
@@ -745,23 +835,22 @@ mempoolCapTestExpectedTrace mcts = go chunks
...
-    chunksOf :: Int -> [a] -> [[a]]
-    chunksOf _ [] = []
-    chunksOf n xs
-      | n > 0     = take n xs : chunksOf n (drop n xs)
-      | otherwise = error $  "mempoolCapTestExpectedTrace.chunksOf: "
-                          <> "n is less than or equal to 0"
+    buildChunks :: [[GenTx TestBlock]] -> [GenTx TestBlock] -> [[GenTx TestBlock]]
+    buildChunks acc []     = acc
+    buildChunks acc remTxs =
+      let (ts, ts') = splitTxsUntilCap remTxs (MempoolCapacity mempoolCap)
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: No need to unwrap the `MempoolCapacity` constructor on L832 and wrap it here again.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 17]
```diff
@@ -745,23 +835,22 @@ mempoolCapTestExpectedTrace mcts = go chunks
...
+      in buildChunks (acc ++ [ts]) ts'
 
     chunks :: [[GenTx TestBlock]]
-    chunks = chunksOf (fromIntegral mempoolCap) txs
+    chunks = buildChunks [] txs
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: You can rewrite `chunks` as follows:
```haskell
chunks []  = []
chunks txs = chunk : chunks txs'
  where
    (chunk, txs') = splitTxsUntilCap txs mempoolCap
```

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 18]
```diff
@@ -286,9 +301,9 @@ prop_Mempool_Capacity mcts = withTestMempool mctsTestSetup $
             -- Indicate that we've removed the transactions from the mempool.
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Actually about L279 (not part of this PR), I would use `unless`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs [hunk 19]
```diff
@@ -276,30 +278,45 @@ data MempoolSnapshot blk idx = MempoolSnapshot {
   , snapshotLookupTx :: idx -> Maybe (GenTx blk)
   }
 
+-- | The size of a mempool.
+data MempoolSize = MempoolSize
+  { msNumTxs   :: !Word
```
[the remark this answers is not in the document]
Comment (intricate) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs]: I suppose I just hadn't switched it to one and left it as a `Word`. However, I think I'd agree with switching this to a `Word32`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs [hunk 20]
```diff
@@ -366,12 +368,24 @@ implAddTxs mpEnv accum txs = assert (all txInvariant txs) $ do
       -- (those not yet validated due to the mempool capacity being reached).
     validateNewUntilMempoolFull []      vr = pure (vr, [])
     validateNewUntilMempoolFull (t:ts)  vr = do
-      mempoolSize <- getMempoolSize mpEnv
+      -- Get the current mempool size
+      mpsz@MempoolSize { msNumBytes } <- getMempoolSize mpEnv
+
+      -- Determine what the mempool size would be if we were to commit the new
+      -- transactions we've validated thus far and also the next transaction
+      -- to validate, 't'.
```
[the remark this answers is not in the document]
Comment (intricate) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs]: Sure, will do.

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs (hunk 8 above)
Comment (intricate) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs]: I agree, I was wondering whether I was the only one that found it a bit unclear.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs (hunk 18 above)
Comment (intricate) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Yeah, that would be nicer than `if ... then pure () else ...`

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs (hunk 9 above)
Comment (intricate) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: Oops. Thanks.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs [hunk 21]
```diff
@@ -715,22 +757,70 @@ instance Arbitrary MempoolCapTestSetup where
...
+-- side would be just enough to fill the mempool to capacity.
+splitTxsUntilCap :: [GenTx TestBlock]
+                 -> MempoolCapacity
+                 -> ([GenTx TestBlock], [GenTx TestBlock])
+splitTxsUntilCap txs (MempoolCapacity mpCap) = go 0 [] txs
+  where
+    go :: Word32
+        -> [GenTx TestBlock]
+        -> [GenTx TestBlock]
+        -> ([GenTx TestBlock], [GenTx TestBlock])
```
Comment (intricate) [inline ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: These are misaligned.

File: ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (intricate) [reply ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs]: Valid point, I agree.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs (hunk 7 above)
Comment (dcoutts) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: A sensible default is 2x the current max block size.

Ultimately we should allow a local override from the node config file. The benchmark team will want this.

File: ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs]: You can now change it to use `splitAfterTxSize`

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs (hunk 7 above)
Comment (intricate) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: For the time being, I've hardcoded this to `128_000` bytes (2x the mainnet max block size).

Comment (intricate) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: Oops, I got this wrong. `128_000` is 2x the mainnet max **transaction** size. From the [mainnet config](https://github.com/input-output-hk/cardano-node/blob/master/configuration/mainnet-ci/configuration.yaml#L292) in `cardano-node`, the max block size appears to be `2_000_000`.
